pkg/runtime/wfengine/workflow.go

```diff
@@ -0,0 +1,245 @@
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		state = &workflowState{
+			inbox: []*backend.HistoryEvent{},
+		}
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+	if exists && !runtimeState.IsCompleted() {
+		return errors.New("workflow instance with this ID already exists")
+	}
+
+	startEvent, err := backend.UnmarshalHistoryEvent(startEventBytes)
+	if err != nil {
+		return err
+	} else if startEvent.GetExecutionStarted() == nil {
+		return errors.New("invalid execution start event")
+	}
+
+	// Schedule a reminder to execute immediately after this operation. The reminder will trigger the actual
+	// workflow execution. This is preferable to using the current thread so that we don't block the client
+	// while the workflow logic is running.
+	if err := wf.createReliableReminder(ctx, actorID, createWorkflowReminder, nil, 0); err != nil {
+		return err
+	}
+
+	state.inbox = []*backend.HistoryEvent{startEvent}
+	return wf.saveInternalState(actorID, state)
+}
+
+func (wf *workflowActor) getWorkflowMetadata(actorID string) (*api.OrchestrationMetadata, error) {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, api.ErrInstanceNotFound
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+
+	name, _ := runtimeState.Name()
+	createdAt, _ := runtimeState.CreatedTime()
+	lastUpdated, _ := runtimeState.LastUpdatedTime()
+	input, _ := runtimeState.Input()
+	output, _ := runtimeState.Output()
+	failureDetuils, _ := runtimeState.FailureDetails()
+
+	metadata := api.NewOrchestrationMetadata(
+		runtimeState.InstanceID(),
+		name,
+		runtimeState.RuntimeStatus(),
+		createdAt,
+		lastUpdated,
+		input,
+		output,
+		runtimeState.CustomStatus.GetValue(),
+		failureDetuils,
+	)
+	return metadata, nil
+}
+
+func (wf *workflowActor) runWorkflow(ctx context.Context, actorID string) error {
+	state, exists, err := wf.loadInternalState(actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		return fmt.Errorf("no workflow state found for actor '%s'", actorID)
+	} else if len(state.inbox) == 0 {
+		// This is never expected - all run requests should be triggered by some inbox event
+		wfLogger.Warnf("%s: ignoring run request for workflow with empty inbox", actorID)
+	}
+
+	runtimeState := getAndCacheRuntimeState(actorID, state)
+	wi := &backend.OrchestrationWorkItem{
+		InstanceID: runtimeState.InstanceID(),
+		NewEvents:  state.inbox,
+		RetryCount: -1, // TODO
+		State:      runtimeState,
+		Properties: make(map[string]interface{}),
+	}
+
+	// Executing workflow code is a one-way operation. We must wait for the app code to report its completion, which
+	// will trigger this callback channel.
+	callback := make(chan bool)
+	wi.Properties[CallbackChannelProperty] = callback
+	wf.scheduler.ScheduleWorkflow(wi)
+	select {
+	case <-ctx.Done():
```

Comment:
Should this `select` intentionally block indefinitely?

Reply:
No, there needs to be a timeout. I added a TODO comment in the code that calls `runWorkflow()` but I can add a similar comment here too to make the intentions more clear.

Reply:
I went ahead and added a 30-second timeout via `context.WithTimeout` (see further up where this method gets called).